### PR TITLE
Refactor rq

### PR DESCRIPTION
**Pull request by micahcc**
Massive pull request refactoring transactions in RQ. I initially made a lot of changes thinking I would fork the library, but I'm now trying to strip away anything not related to the transaction / multi dependency improvements.

Changes:
- Created RQConnection so that we don't have to patch redis connections
- RQConnection also stores the worker_class, queue_class and job_class, now it should be possible to have the same custom class everywhere instead of just in a queue or worker, which definitely could cause issues
- pipes aren't explicitly passed around, the pipe is stored in RQConnection as _pipe and transaction decorator uses property `rq_conn` to initialize pipes. Not beautiful but it is pretty foolproof; much easier to ensure that pipes are used consistently.
- dependents and dependencies have been changed to parents and children. Its too frigging hard to know what a dependent is otherwise. Parents come before, children come after.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,3 +1,10 @@
+## Implementation / Refactor
+Replace old patched redis function with a new Storage object.
+
+- Storage object *requires* everything to be a transaction
+- Storage object handles popping correctly (read then remove)
```
Comment (leighmcculloch) [inline README.md]: 👍

File: rq/cli/__init__.py [hunk 2]
```diff
@@ -4,3 +4,5 @@
 # TODO: the following imports can be removed when we drop the `rqinfo` and
 # `rqworkers` commands in favor of just shipping the `rq` command.
 from .cli import info, worker
+
+
```
Comment (leighmcculloch) [inline rq/cli/__init__.py]: ?

File: rq/connections.py [hunk 3]
```diff
@@ -1,36 +1,479 @@
...
-    push_connection(connection)
-    try:
-        yield
-    finally:
-        popped = pop_connection()
-        assert popped == connection, \
-            'Unexpected Redis connection was popped off the stack. ' \
-            'Check your Redis connection setup.'
+class RQConnection(object):
+    from .worker import Worker
```
Comment (proofit404) [inline rq/connections.py]: Use top level imports.

Comment (micahcc) [inline rq/connections.py]: I was worried about circular imports. I'll check to see if its possible. The issue is Worker definitely should know about connection, and RQConnection stores a worker class. Maybe the trick is moving push/pop connection to another module, but somewhere we have to break the cycle. Thanks for commenting

Comment (proofit404) [inline rq/connections.py]: This class statements executed during module import. So moving your import doesn't change the circle. You can define worker as a string in the class attribute

``` python
class RQConnection:
    worker_class = 'rq.worker.Worker'
```

Then use `import_attribute` from `rq.utils` similar as job deserialization do.

Comment (micahcc) [inline rq/connections.py]: Ahh good point.

**Pull request merged**